doc/cookbook/current-route/CurrentRoute.lhs

```diff
@@ -0,0 +1,391 @@
+
+```haskell
+data Routes4 route = Routes4
+    { route41 :: route :- "route1" :> Get '[JSON] Int
+    , route42 :: route :- "route2" :> Capture "id" Int :> Get '[JSON] Int
+    }
+  deriving (Generic)
+```
+
+But instead of making a `Routes4 AsServer` value, we'll use a different *mode*.
+`AsRecServer route` is a handler which takes `route :- api` as a first
+argument.  In this example we use `HasLink'`, but reader is free to use other
+automatic interpretations, e.g. `servant-client` to make a proxy!
+
+```haskell
+data AsRecServer route
+instance GenericMode (AsRecServer route) where
+    type AsRecServer route :- api = (route :- api) -> (AsServer :- api)
+
+routes4 :: Routes4 (AsRecServer (AsLink Link))
+routes4 = Routes4
+    { route41 = \l -> liftIO $ do
+        print l
+        return 42
+    , route42 = \l i -> liftIO $ do
+        print (l i)
+        return i
+    }
+
+app4 :: Application
+app4 = genericRecServe routes4
+```
+
+The usage is very simple, unfortunately the implementation is not.
+
+### Hairy bits
+
+The implementation of `genericRecServe` is intimidating.
+The missing bit is a function `genericHoist`.
+In short, given a function which can convert `modeA :- api` into `modeB :- api` for all `api`,
+`genericHoist` converts `routes modeA` into `routes modeB`.
+Maybe this function should exist in `Servant.API.Generic`?
+
+```haskell
+genericHoist
+    :: ( GenericMode modeA, GenericMode modeB
+       , Generic (routes modeA), Generic (routes modeB)
+       , GServantHoist c api modeA modeB (Rep (routes modeA)) (Rep (routes modeB))
+       )
+    => Proxy modeA -> Proxy modeB -> Proxy c -> Proxy api
+    -> (forall api'. c api' => Proxy api' -> (modeA :- api') -> (modeB :- api'))
+    -> routes modeA -> routes modeB
+genericHoist pa pb pc api nt = to . gservantHoist pa pb pc api nt . from
+```
+
+`genericRecServe` is `genericHoist` precomposed with a variant of `genericServe`.
+The implementation of one-liner, given a wall of constraints.
+
+```haskell
+genericRecServe
+    :: forall routes.
+       ( HasServer (ToServantApi routes) '[]
+       , GenericServant routes AsApi
+       , GenericServant routes AsServer
+       , GenericServant routes (AsRecServer (AsLink Link))
+       , Server (ToServantApi routes) ~ ToServant routes AsServer
+       , GServantHoist 
+          HasLink'
+          (ToServantApi routes)
+          (AsRecServer (AsLink Link))
+          AsServer
+          (Rep (routes (AsRecServer (AsLink Link))))
+          (Rep (routes AsServer))
+       )
+    => routes (AsRecServer (AsLink Link)) -> Application
+genericRecServe
+    = serve (Proxy :: Proxy (ToServantApi routes)) 
+    . toServant
+    . genericHoist
+        (Proxy :: Proxy (AsRecServer (AsLink Link)))
+        (Proxy :: Proxy AsServer)
+        (Proxy :: Proxy HasLink')
+        (genericApi (Proxy :: Proxy routes))
+        (\p f -> f $ safeLink p p)
+```
+
+There we us single-instance-class trick to make partially applicable `HasLink`.
```

Comment:
Do you have a link we could give at this point to help the reader learn about it?

Reply:
Let's see if we can find one